### ISSUE TITLE
Crash "assertion failed: b.tv_nsec >= 0" in gRPC 1.19

### ISSUE THREAD

**Issue by mehrdada** [grpc/grpc#18244]
We are seeing a new crash that manifests itself in 1.19.0. Seems like this PR is the only thing that touched it in a year and the assertion seems failing: https://github.com/grpc/grpc/pull/17842 

(We are using Python but seems like the problem is probably not Python specific.) This is blocking us from upgrading from 1.18.

This is the tail of the log `GRPC_VERBOSITY=debug GRPC_TRACE=all`:
```
D0304 21:59:56.199154892    2043 transport.cc:45]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910   REF 5->6 external_connectivity_watcher
I0304 21:59:56.199157956    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c01f170 last=1
I0304 21:59:56.199161265    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199164048    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c01f170
D0304 21:59:56.199167989    2043 exec_ctx.cc:34]             running closure 0x7f191c014780: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:239]: run [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2688]
I0304 21:59:56.199172121    2043 timer_generic.cc:365]       TIMER 0x7f191c0147d0: SET 0 now 426 call 0x7f191c014730[0x7f1945a84683]
D0304 21:59:56.199175259    2043 exec_ctx.cc:43]             closure 0x7f191c014780 finished
I0304 21:59:56.199178806    2043 connectivity_state.cc:116]  CONWATCH: 0x7f1918009ae0 request_router: from TRANSIENT_FAILURE [cur=TRANSIENT_FAILURE] notify=0x7f191c01f170
I0304 21:59:56.199182556    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=3
D0304 21:59:56.199187173    2043 exec_ctx.cc:34]             running closure 0x7f191c014730: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:226]: scheduled [external/com_github_grpc_grpc/src/core/lib/iomgr/timer_generic.cc:383]
I0304 21:59:56.199192041    2043 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.199195997    2043 ev_epollex_linux.cc:1459]   PSS:0x7f1918002fd0: add pollset 0x7f191800acf8
D0304 21:59:56.199199917    2043 ev_epollex_linux.cc:1250]   POLLABLE:0x7f191c014860   ref 1->2 pollset_as_multipollable
D0304 21:59:56.199203373    2043 ev_epollex_linux.cc:1289]   POLLABLE:0x7f191c014860   ref 2->3 pollset_set
D0304 21:59:56.199206968    2043 ev_epollex_linux.cc:1290]   POLLABLE:0x7f191c014860 unref 3->2 pollset_as_multipollable
D0304 21:59:56.199211169    2043 ev_epollex_linux.cc:1487]   POLLABLE:0x7f191c014860 unref 2->1 pollset_set
D0304 21:59:56.199214982    2043 transport.cc:45]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910   REF 6->7 external_connectivity_watcher
I0304 21:59:56.199218800    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c01f960 last=1
D0304 21:59:56.199223424    2043 error.cc:337]               0x7f191c01fbc0 create [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:136]
D0304 21:59:56.199228267    2043 error.cc:127]               0x7f191c01fbc0: 1 -> 2 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:143]
D0304 21:59:56.199232891    2043 error.cc:175]               0x7f191c01fbc0: 2 -> 1 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:164]
D0304 21:59:56.199236700    2043 exec_ctx.cc:43]             closure 0x7f191c014730 finished
I0304 21:59:56.199240775    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199244883    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c01f960
I0304 21:59:56.199248979    2043 connectivity_state.cc:113]  CONWATCH: 0x7f1918009ae0 request_router: unsubscribe notify=0x7f191c01f170
I0304 21:59:56.199252866    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c01f170 last=3
I0304 21:59:56.199256898    2043 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.199260247    2043 ev_epollex_linux.cc:1407]   PSS:0x7f1918002fd0: del pollset 0x7f191800acf8
D0304 21:59:56.199263844    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 7->6 external_connectivity_watcher
I0304 21:59:56.199267157    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=5
I0304 21:59:56.199270971    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199273875    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c01f170
I0304 21:59:56.199276869    2043 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.199280283    2043 ev_epollex_linux.cc:1407]   PSS:0x7f1918002fd0: del pollset 0x7f191800acf8
I0304 21:59:56.199284900    2043 ev_epollex_linux.cc:674]    PS:0x7f191800acf8 (pollable:0x7f191c014860) maybe_finish_shutdown sc=(nil) (target:!NULL) rw=(nil) (target:NULL) cpsc=0 (target:0)
D0304 21:59:56.199289836    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 6->5 external_connectivity_watcher
I0304 21:59:56.199293946    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=3
D0304 21:59:56.199298625    2043 exec_ctx.cc:34]             running closure 0x7f191c0146e0: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:224]: scheduled [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2676]
I0304 21:59:56.199303379    2043 timer_generic.cc:467]       TIMER 0x7f191c0147d0: CANCEL pending=false
E0304 21:59:56.199306523    2043 channel_connectivity.cc:129] watch_completion_error: "Cancelled"
I0304 21:59:56.199320462    2043 completion_queue.cc:699]    cq_end_op_for_next(cq=0x7f191800abf0, tag=0x7f193f3e7b90, error={"created":"@1551736796.199226674","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}, done=0x7f1945a842c3, done_arg=0x7f191c0146b0, storage=0x7f191c014818)
E0304 21:59:56.199329413    2043 completion_queue.cc:701]    Operation failed: tag=0x7f193f3e7b90, error={"created":"@1551736796.199226674","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}
I0304 21:59:56.199333695    2043 ev_posix.cc:329]            (polling-api) pollset_kick(0x7f191800acf8, (nil))
I0304 21:59:56.199336607    2043 ev_epollex_linux.cc:740]    PS:0x7f191800acf8 kick (nil) tls_pollset=(nil) tls_worker=(nil) pollset.root_worker=(nil)
I0304 21:59:56.199339354    2043 ev_epollex_linux.cc:750]    PS:0x7f191800acf8 kicked_any_without_poller
D0304 21:59:56.199342523    2043 error.cc:175]               0x7f191c01fbc0: 1 -> 0 [external/com_github_grpc_grpc/src/core/lib/surface/completion_queue.cc:760]
D0304 21:59:56.199345848    2043 exec_ctx.cc:43]             closure 0x7f191c0146e0 finished
I0304 21:59:56.199360457    2043 completion_queue.cc:959]    grpc_completion_queue_next(cq=0x7f191800abf0, deadline=gpr_timespec { tv_sec: 1551736796, tv_nsec: 399358724, clock_type: 1 }, reserved=(nil))
D0304 21:59:56.199365028    2043 completion_queue.cc:937]    PENDING TAGS:
D0304 21:59:56.199368654    2043 completion_queue.cc:964]    CQ:0x7f191800abf0   ref 2 -> 3 next
D0304 21:59:56.199373950    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 5->4 watch_channel_connectivity
I0304 21:59:56.199379502    2043 completion_queue.cc:1059]   RETURN_EVENT[0x7f191800abf0]: OP_COMPLETE: tag:0x7f193f3e7b90 ERROR
D0304 21:59:56.199383740    2043 completion_queue.cc:1060]   CQ:0x7f191800abf0 unref 3 -> 2 next
I0304 21:59:56.199429836    2045 channel_connectivity.cc:218] grpc_channel_watch_connectivity_state(channel=0x793ab60, last_observed_state=3, deadline=gpr_timespec { tv_sec: -62118953636, tv_nsec: -800003051, clock_type: 1 }, cq=0x793b900, tag=0x7f193f3e7850)
D0304 21:59:56.199442850    2045 transport.cc:45]            CLIENT_CHANNEL 0x793abc8:0x793ab60   REF 4->5 watch_channel_connectivity
I0304 21:59:56.199447820    2045 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7937af0, 0x793ba08)
I0304 21:59:56.199451806    2045 ev_epollex_linux.cc:1459]   PSS:0x7937af0: add pollset 0x793ba08
D0304 21:59:56.199457001    2045 ev_epollex_linux.cc:1250]   POLLABLE:0x7f1914014660   ref 1->2 pollset_as_multipollable
D0304 21:59:56.199461686    2045 ev_epollex_linux.cc:1289]   POLLABLE:0x7f1914014660   ref 2->3 pollset_set
D0304 21:59:56.199465486    2045 ev_epollex_linux.cc:1290]   POLLABLE:0x7f1914014660 unref 3->2 pollset_as_multipollable
D0304 21:59:56.199469584    2045 ev_epollex_linux.cc:1487]   POLLABLE:0x7f1914014660 unref 2->1 pollset_set
D0304 21:59:56.199473875    2045 transport.cc:45]            CLIENT_CHANNEL 0x793abc8:0x793ab60   REF 5->6 external_connectivity_watcher
I0304 21:59:56.199478668    2045 combiner.cc:167]            C:0x793ae00 grpc_combiner_execute c=0x7f191401c7d0 last=1
I0304 21:59:56.199483184    2045 combiner.cc:231]            C:0x793ae00 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199487575    2045 combiner.cc:254]            C:0x793ae00 maybe_finish_one n=0x7f191401c7d0
D0304 21:59:56.199493261    2045 exec_ctx.cc:34]             running closure 0x7f19140176a0: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:239]: run [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2688]
I0304 21:59:56.199499542    2045 timer_generic.cc:365]       TIMER 0x7f19140176f0: SET 0 now 426 call 0x7f1914017650[0x7f1945a84683]
D0304 21:59:56.199504160    2045 exec_ctx.cc:43]             closure 0x7f19140176a0 finished
I0304 21:59:56.199509314    2045 connectivity_state.cc:116]  CONWATCH: 0x793ad30 request_router: from TRANSIENT_FAILURE [cur=TRANSIENT_FAILURE] notify=0x7f191401c7d0
I0304 21:59:56.199513837    2045 combiner.cc:296]            C:0x793ae00 finish old_state=3
D0304 21:59:56.199518469    2045 exec_ctx.cc:34]             running closure 0x7f1914017650: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:226]: scheduled [external/com_github_grpc_grpc/src/core/lib/iomgr/timer_generic.cc:383]
I0304 21:59:56.199523572    2045 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7937af0, 0x793ba08)
I0304 21:59:56.199527686    2045 ev_epollex_linux.cc:1459]   PSS:0x7937af0: add pollset 0x793ba08
D0304 21:59:56.199532003    2045 ev_epollex_linux.cc:1250]   POLLABLE:0x7f1914014660   ref 1->2 pollset_as_multipollable
D0304 21:59:56.199535943    2045 ev_epollex_linux.cc:1289]   POLLABLE:0x7f1914014660   ref 2->3 pollset_set
D0304 21:59:56.199539838    2045 ev_epollex_linux.cc:1290]   POLLABLE:0x7f1914014660 unref 3->2 pollset_as_multipollable
D0304 21:59:56.199543991    2045 ev_epollex_linux.cc:1487]   POLLABLE:0x7f1914014660 unref 2->1 pollset_set
D0304 21:59:56.199548428    2045 transport.cc:45]            CLIENT_CHANNEL 0x793abc8:0x793ab60   REF 6->7 external_connectivity_watcher
I0304 21:59:56.199552340    2045 combiner.cc:167]            C:0x793ae00 grpc_combiner_execute c=0x7f191401cf40 last=1
D0304 21:59:56.199557588    2045 error.cc:337]               0x7f191401d0c0 create [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:136]
D0304 21:59:56.199563319    2045 error.cc:127]               0x7f191401d0c0: 1 -> 2 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:143]
D0304 21:59:56.199568611    2045 error.cc:175]               0x7f191401d0c0: 2 -> 1 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:164]
D0304 21:59:56.199572987    2045 exec_ctx.cc:43]             closure 0x7f1914017650 finished
I0304 21:59:56.199577207    2045 combiner.cc:231]            C:0x793ae00 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199581519    2045 combiner.cc:254]            C:0x793ae00 maybe_finish_one n=0x7f191401cf40
I0304 21:59:56.199585502    2045 connectivity_state.cc:113]  CONWATCH: 0x793ad30 request_router: unsubscribe notify=0x7f191401c7d0
I0304 21:59:56.199589932    2045 combiner.cc:167]            C:0x793ae00 grpc_combiner_execute c=0x7f191401c7d0 last=3
I0304 21:59:56.199593930    2045 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7937af0, 0x793ba08)
I0304 21:59:56.199597679    2045 ev_epollex_linux.cc:1407]   PSS:0x7937af0: del pollset 0x793ba08
D0304 21:59:56.199602106    2045 transport.cc:59]            CLIENT_CHANNEL 0x793abc8:0x793ab60 UNREF 7->6 external_connectivity_watcher
I0304 21:59:56.199606119    2045 combiner.cc:296]            C:0x793ae00 finish old_state=5
I0304 21:59:56.199610390    2045 combiner.cc:231]            C:0x793ae00 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199614059    2045 combiner.cc:254]            C:0x793ae00 maybe_finish_one n=0x7f191401c7d0
I0304 21:59:56.199617552    2045 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7937af0, 0x793ba08)
I0304 21:59:56.199621072    2045 ev_epollex_linux.cc:1407]   PSS:0x7937af0: del pollset 0x793ba08
I0304 21:59:56.199626168    2045 ev_epollex_linux.cc:674]    PS:0x793ba08 (pollable:0x7f1914014660) maybe_finish_shutdown sc=(nil) (target:!NULL) rw=(nil) (target:NULL) cpsc=0 (target:0)
D0304 21:59:56.199630990    2045 transport.cc:59]            CLIENT_CHANNEL 0x793abc8:0x793ab60 UNREF 6->5 external_connectivity_watcher
I0304 21:59:56.199635172    2045 combiner.cc:296]            C:0x793ae00 finish old_state=3
D0304 21:59:56.199640203    2045 exec_ctx.cc:34]             running closure 0x7f1914017600: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:224]: scheduled [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2676]
I0304 21:59:56.199645333    2045 timer_generic.cc:467]       TIMER 0x7f19140176f0: CANCEL pending=false
E0304 21:59:56.199649691    2045 channel_connectivity.cc:129] watch_completion_error: "Cancelled"
I0304 21:59:56.199666081    2045 completion_queue.cc:699]    cq_end_op_for_next(cq=0x793b900, tag=0x7f193f3e7850, error={"created":"@1551736796.199561402","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}, done=0x7f1945a842c3, done_arg=0x7f19140175d0, storage=0x7f1914017738)
E0304 21:59:56.199674517    2045 completion_queue.cc:701]    Operation failed: tag=0x7f193f3e7850, error={"created":"@1551736796.199561402","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}
I0304 21:59:56.199680469    2045 ev_posix.cc:329]            (polling-api) pollset_kick(0x793ba08, (nil))
I0304 21:59:56.199684662    2045 ev_epollex_linux.cc:740]    PS:0x793ba08 kick (nil) tls_pollset=(nil) tls_worker=(nil) pollset.root_worker=(nil)
I0304 21:59:56.199688874    2045 ev_epollex_linux.cc:750]    PS:0x793ba08 kicked_any_without_poller
D0304 21:59:56.199693557    2045 error.cc:175]               0x7f191401d0c0: 1 -> 0 [external/com_github_grpc_grpc/src/core/lib/surface/completion_queue.cc:760]
D0304 21:59:56.199698122    2045 exec_ctx.cc:43]             closure 0x7f1914017600 finished
I0304 21:59:56.199715291    2045 completion_queue.cc:959]    grpc_completion_queue_next(cq=0x793b900, deadline=gpr_timespec { tv_sec: 1551736796, tv_nsec: 399713692, clock_type: 1 }, reserved=(nil))
D0304 21:59:56.199724548    2045 completion_queue.cc:937]    PENDING TAGS:
D0304 21:59:56.199728276    2045 completion_queue.cc:964]    CQ:0x793b900   ref 2 -> 3 next
D0304 21:59:56.199733158    2045 transport.cc:59]            CLIENT_CHANNEL 0x793abc8:0x793ab60 UNREF 5->4 watch_channel_connectivity
I0304 21:59:56.199738888    2045 completion_queue.cc:1059]   RETURN_EVENT[0x793b900]: OP_COMPLETE: tag:0x7f193f3e7850 ERROR
D0304 21:59:56.199743150    2045 completion_queue.cc:1060]   CQ:0x793b900 unref 3 -> 2 next
I0304 21:59:56.199839665    2043 channel_connectivity.cc:218] grpc_channel_watch_connectivity_state(channel=0x7f1918009910, last_observed_state=3, deadline=gpr_timespec { tv_sec: -62118953636, tv_nsec: -800003051, clock_type: 1 }, cq=0x7f191800abf0, tag=0x7f193f3e7e10)
D0304 21:59:56.199854856    2043 transport.cc:45]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910   REF 4->5 watch_channel_connectivity
I0304 21:59:56.199860962    2043 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.199865675    2043 ev_epollex_linux.cc:1459]   PSS:0x7f1918002fd0: add pollset 0x7f191800acf8
D0304 21:59:56.199876338    2043 ev_epollex_linux.cc:1250]   POLLABLE:0x7f191c014860   ref 1->2 pollset_as_multipollable
D0304 21:59:56.199880823    2043 ev_epollex_linux.cc:1289]   POLLABLE:0x7f191c014860   ref 2->3 pollset_set
D0304 21:59:56.199885075    2043 ev_epollex_linux.cc:1290]   POLLABLE:0x7f191c014860 unref 3->2 pollset_as_multipollable
D0304 21:59:56.199889718    2043 ev_epollex_linux.cc:1487]   POLLABLE:0x7f191c014860 unref 2->1 pollset_set
D0304 21:59:56.199892941    2043 transport.cc:45]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910   REF 5->6 external_connectivity_watcher
I0304 21:59:56.199896291    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c01fbf0 last=1
I0304 21:59:56.199900551    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199903397    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c01fbf0
D0304 21:59:56.199907718    2043 exec_ctx.cc:34]             running closure 0x7f191c014780: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:239]: run [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2688]
I0304 21:59:56.199919273    2043 timer_generic.cc:365]       TIMER 0x7f191c0147d0: SET 0 now 426 call 0x7f191c014730[0x7f1945a84683]
D0304 21:59:56.199923728    2043 exec_ctx.cc:43]             closure 0x7f191c014780 finished
I0304 21:59:56.199928998    2043 connectivity_state.cc:116]  CONWATCH: 0x7f1918009ae0 request_router: from TRANSIENT_FAILURE [cur=TRANSIENT_FAILURE] notify=0x7f191c01fbf0
I0304 21:59:56.199933805    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=3
D0304 21:59:56.199938480    2043 exec_ctx.cc:34]             running closure 0x7f191c014730: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:226]: scheduled [external/com_github_grpc_grpc/src/core/lib/iomgr/timer_generic.cc:383]
I0304 21:59:56.199943431    2043 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.199947618    2043 ev_epollex_linux.cc:1459]   PSS:0x7f1918002fd0: add pollset 0x7f191800acf8
D0304 21:59:56.199951612    2043 ev_epollex_linux.cc:1250]   POLLABLE:0x7f191c014860   ref 1->2 pollset_as_multipollable
D0304 21:59:56.199955806    2043 ev_epollex_linux.cc:1289]   POLLABLE:0x7f191c014860   ref 2->3 pollset_set
D0304 21:59:56.199959756    2043 ev_epollex_linux.cc:1290]   POLLABLE:0x7f191c014860 unref 3->2 pollset_as_multipollable
D0304 21:59:56.199963989    2043 ev_epollex_linux.cc:1487]   POLLABLE:0x7f191c014860 unref 2->1 pollset_set
D0304 21:59:56.199967976    2043 transport.cc:45]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910   REF 6->7 external_connectivity_watcher
I0304 21:59:56.199971769    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c0203e0 last=1
D0304 21:59:56.199977097    2043 error.cc:337]               0x7f191c020640 create [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:136]
D0304 21:59:56.199982119    2043 error.cc:127]               0x7f191c020640: 1 -> 2 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:143]
D0304 21:59:56.199986614    2043 error.cc:175]               0x7f191c020640: 2 -> 1 [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:164]
D0304 21:59:56.199990531    2043 exec_ctx.cc:43]             closure 0x7f191c014730 finished
I0304 21:59:56.199995608    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.199999526    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c0203e0
I0304 21:59:56.200003674    2043 connectivity_state.cc:113]  CONWATCH: 0x7f1918009ae0 request_router: unsubscribe notify=0x7f191c01fbf0
I0304 21:59:56.200006681    2043 combiner.cc:167]            C:0x7f1918009490 grpc_combiner_execute c=0x7f191c01fbf0 last=3
I0304 21:59:56.200009555    2043 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.200012921    2043 ev_epollex_linux.cc:1407]   PSS:0x7f1918002fd0: del pollset 0x7f191800acf8
D0304 21:59:56.200017063    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 7->6 external_connectivity_watcher
I0304 21:59:56.200020965    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=5
I0304 21:59:56.200025042    2043 combiner.cc:231]            C:0x7f1918009490 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.200028618    2043 combiner.cc:254]            C:0x7f1918009490 maybe_finish_one n=0x7f191c01fbf0
I0304 21:59:56.200032340    2043 ev_posix.cc:369]            (polling-api) pollset_set_del_pollset(0x7f1918002fd0, 0x7f191800acf8)
I0304 21:59:56.200035926    2043 ev_epollex_linux.cc:1407]   PSS:0x7f1918002fd0: del pollset 0x7f191800acf8
I0304 21:59:56.200040995    2043 ev_epollex_linux.cc:674]    PS:0x7f191800acf8 (pollable:0x7f191c014860) maybe_finish_shutdown sc=(nil) (target:!NULL) rw=(nil) (target:NULL) cpsc=0 (target:0)
D0304 21:59:56.200045958    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 6->5 external_connectivity_watcher
I0304 21:59:56.200049985    2043 combiner.cc:296]            C:0x7f1918009490 finish old_state=3
D0304 21:59:56.200054785    2043 exec_ctx.cc:34]             running closure 0x7f191c0146e0: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:224]: scheduled [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2676]
I0304 21:59:56.200059661    2043 timer_generic.cc:467]       TIMER 0x7f191c0147d0: CANCEL pending=false
E0304 21:59:56.200072311    2043 channel_connectivity.cc:129] watch_completion_error: "Cancelled"
I0304 21:59:56.200089684    2043 completion_queue.cc:699]    cq_end_op_for_next(cq=0x7f191800abf0, tag=0x7f193f3e7e10, error={"created":"@1551736796.199980708","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}, done=0x7f1945a842c3, done_arg=0x7f191c0146b0, storage=0x7f191c014818)
E0304 21:59:56.200108294    2043 completion_queue.cc:701]    Operation failed: tag=0x7f193f3e7e10, error={"created":"@1551736796.199980708","description":"Timed out waiting for connection state change","file":"external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc","file_line":136}
I0304 21:59:56.200114528    2043 ev_posix.cc:329]            (polling-api) pollset_kick(0x7f191800acf8, (nil))
I0304 21:59:56.200126988    2043 ev_epollex_linux.cc:740]    PS:0x7f191800acf8 kick (nil) tls_pollset=(nil) tls_worker=(nil) pollset.root_worker=(nil)
I0304 21:59:56.200129993    2043 ev_epollex_linux.cc:750]    PS:0x7f191800acf8 kicked_any_without_poller
D0304 21:59:56.200133584    2043 error.cc:175]               0x7f191c020640: 1 -> 0 [external/com_github_grpc_grpc/src/core/lib/surface/completion_queue.cc:760]
D0304 21:59:56.200136823    2043 exec_ctx.cc:43]             closure 0x7f191c0146e0 finished
I0304 21:59:56.200164937    2043 completion_queue.cc:959]    grpc_completion_queue_next(cq=0x7f191800abf0, deadline=gpr_timespec { tv_sec: 1551736796, tv_nsec: 400163153, clock_type: 1 }, reserved=(nil))
D0304 21:59:56.200175430    2043 completion_queue.cc:937]    PENDING TAGS:
D0304 21:59:56.200179322    2043 completion_queue.cc:964]    CQ:0x7f191800abf0   ref 2 -> 3 next
D0304 21:59:56.200184926    2043 transport.cc:59]            CLIENT_CHANNEL 0x7f1918009978:0x7f1918009910 UNREF 5->4 watch_channel_connectivity
I0304 21:59:56.200191667    2043 completion_queue.cc:1059]   RETURN_EVENT[0x7f191800abf0]: OP_COMPLETE: tag:0x7f193f3e7e10 ERROR
D0304 21:59:56.200196674    2043 completion_queue.cc:1060]   CQ:0x7f191800abf0 unref 3 -> 2 next
I0304 21:59:56.200316731    2045 channel_connectivity.cc:218] grpc_channel_watch_connectivity_state(channel=0x793ab60, last_observed_state=3, deadline=gpr_timespec { tv_sec: -62118953636, tv_nsec: -800003051, clock_type: 1 }, cq=0x793b900, tag=0x7f193f3e7f50)
D0304 21:59:56.200329818    2045 transport.cc:45]            CLIENT_CHANNEL 0x793abc8:0x793ab60   REF 4->5 watch_channel_connectivity
I0304 21:59:56.200335176    2045 ev_posix.cc:362]            (polling-api) pollset_set_add_pollset(0x7937af0, 0x793ba08)
I0304 21:59:56.200339358    2045 ev_epollex_linux.cc:1459]   PSS:0x7937af0: add pollset 0x793ba08
D0304 21:59:56.200343890    2045 ev_epollex_linux.cc:1250]   POLLABLE:0x7f1914014660   ref 1->2 pollset_as_multipollable
D0304 21:59:56.200348718    2045 ev_epollex_linux.cc:1289]   POLLABLE:0x7f1914014660   ref 2->3 pollset_set
D0304 21:59:56.200351417    2045 ev_epollex_linux.cc:1290]   POLLABLE:0x7f1914014660 unref 3->2 pollset_as_multipollable
D0304 21:59:56.200354627    2045 ev_epollex_linux.cc:1487]   POLLABLE:0x7f1914014660 unref 2->1 pollset_set
D0304 21:59:56.200358263    2045 transport.cc:45]            CLIENT_CHANNEL 0x793abc8:0x793ab60   REF 5->6 external_connectivity_watcher
I0304 21:59:56.200363018    2045 combiner.cc:167]            C:0x793ae00 grpc_combiner_execute c=0x7f191401d0f0 last=1
I0304 21:59:56.200367951    2045 combiner.cc:231]            C:0x793ae00 grpc_combiner_continue_exec_ctx contended=0 exec_ctx_ready_to_finish=1 time_to_execute_final_list=0
I0304 21:59:56.200372349    2045 combiner.cc:254]            C:0x793ae00 maybe_finish_one n=0x7f191401d0f0
D0304 21:59:56.200378420    2045 exec_ctx.cc:34]             running closure 0x7f19140176a0: created [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:239]: run [external/com_github_grpc_grpc/src/core/ext/filters/client_channel/client_channel.cc:2688]
E0304 21:59:56.200383299    2045 time.cc:141]                assertion failed: b.tv_nsec >= 0
```

**Comment by mehrdada** [comment]
cc @muxi @rmstar

**Comment by srini100** [comment]
cc @yang-g

**Comment by mehrdada** [comment]
Upon further investigation, it seems out test (big test) uses `freezegun.freeze_time` module and that is probably interacting weirdly with the Python/Cython layer.

**Comment by mehrdada** [comment]
@vjpai @srini100 I consider `freeze_time` a shady library so this might not warrant a "p1" label, as long as the assertions make sense.

**Comment by rmstar** [comment]
I believe the assert is correct: tv_nsec should always be >= 0. The assert seems to be exposing an issue in C-Core where we somehow end up with negative tv_nsec

**Comment by yang-g** [comment]
Yes. It should be a number between `[0, 1e9-1]`. I believe c core tries to keep it in range. If we see this assert is failing, we should look into the details. I believe there are issues in ios where the number is out of range and caused production failure. But I am not sure whether we figured that out or not.

**Comment by rmstar** [comment]
@yang-g the issues seen on ios have not been root caused yet. See internal bug b/113140355 for an example.

**Comment by mehrdada** [comment]
I think this is the root cause: https://github.com/grpc/grpc/blob/c339ba32298a00436ba98800d513dcf4d0944d11/src/python/grpcio/grpc/_cython/_cygrpc/time.pyx.pxi#L22

**Comment by yang-g** [comment]
@rmstar understood. I think it is most likely caused by some issue related to tv_nsec being negative and thus we added additional debug logging..

**Comment by srini100** [comment]
@mehrdada, can you please confirm if your PR fixes your internal test so we can roll it into a patch?

**Comment by mehrdada** [comment]
@srini100 Thanks for the help. That PR is still in progress and @lidizheng is helping with the Windows test failures. We changed our test timestamp to not run into this issue, so we are no longer blocked by this. If there are no other symptoms outside Python that you have seen, this should be python-specific and a low priority one at that and I don't think warrants a point release. We've upgraded our gRPC core and seems fine so far from other perspectives.
